src/core/Akka.Persistence/Eventsourced.cs

```diff
@@ -501,6 +497,52 @@ private void StashInternally(object currentMessage)
                 }
             }
         }
+
+        private class InternalStashAwareStash : IStash
+        {
+            private readonly IStash _userStash;
+            private readonly IStash _internalStash;
+
+            public InternalStashAwareStash(IStash userStash, IStash internalStash)
+            {
+                _userStash = userStash;
+                _internalStash = internalStash;
+            }
+
+            public void Stash()
+            {
+                _userStash.Stash();
+            }
+
+            public void Unstash()
+            {
+                _userStash.Unstash();
+            }
+
+            public void UnstashAll()
+            {
+                // Internally, all messages are processed by unstashing them from
+                // the internal stash one-by-one. Hence, an unstashAll() from the
+                // user stash must be prepended to the internal stash.
+
+                _internalStash.Prepend(ClearStash());
+            }
+
+            public void UnstashAll(Func<Envelope, bool> predicate)
```

Comment:
Isn't this a little confusing? I mean, this stash consists of two internally, and while parameterless `UnstashAll` method is working on internal stash, the one with predicate works only with user stash. Maybe we should name them in a way, that is more informative about their purpose.